XMonad/Hooks/EwmhDesktops.hs

```diff
@@ -320,10 +320,28 @@ ewmhDesktopsLogHook' EwmhDesktopsConfig{workspaceSort, workspaceRename} = withWi
     let clientList = nub . concatMap (W.integrate' . W.stack) $ ws
     whenChanged (ClientList clientList) $ setClientList clientList
 
-    -- Set stacking client list which should have bottom-to-top
-    -- stacking order, i.e. focused window should be last
-    let clientListStacking = nub . concatMap (maybe [] (\(W.Stack x l r) -> reverse l ++ r ++ [x]) . W.stack) $ ws
-    whenChanged (ClientListStacking clientListStacking) $ setClientListStacking clientListStacking
+    -- @ws@ is sorted in the "workspace order", which, by default, is
+    -- the lexicographical sorting on @WorkspaceId@.
+    -- @_NET_CLIENT_LIST_STACKING@ is expected to be in the "bottom-to-top
+    -- stacking order".  It is unclear what that would mean for windows on
+    -- invisible workspaces, but it seems reasonable to assume that windows on
+    -- the current workspace should be "at the top".  With the focused window to
+    -- be the top most, meaning the last.
+    let clientListStacking =
+          let wsInFocusOrder = W.hidden s
+                               ++ (map W.workspace . W.visible) s
+                               ++ [W.workspace $ W.current s]
+              stackWindows (W.Stack cur up down) = down ++ reverse up ++ [cur]
```

Comment:
Based on the previous code, shouldn't this be like this? Or was there a mistake there too?
```haskell
              stackWindows (W.Stack cur up down) = reverse up ++ down ++ [cur]
```

Reply:
When I was writing this, I thought that the previous code incorrect.
But now that you ask, I am not so sure :)

"Bottom-to-top focus order" means that things that are higher in the focus order should be closer to the end of the list.
But I actually do not know if `up` or `down` is related to the visual stacking.
I think it depends on the layout used for the workspace.
It seems reasonable to assume that either the first element of `up` (because it is reversed) or the first element of `down` would be next to `cur` in the focus order?

Meaning that it could be either of:
```haskell
stackWindows (W.Stack cur up down) = reverse down ++ up ++ [cur]
```
or
```haskell
stackWindows (W.Stack cur up down) = up ++ reverse down ++ [cur]
```

What do you think?

Reply:
The problem with this is that in a tiling window manager there is no actual "visual stacking", and in xmonad the usual focus changing primitives use `up` or `down` (pushing `focus` onto one and then popping the other into `focus`, unless this causes wraparound).

Reply:
Also, for those layouts that don't actually tile as such, there's no guarantee that what the layout does matches what the window stack suggests, and it might be necessary to duplicate `windows`'s zipping the stack against the result of `runLayout` (a list of rectangles in hopefully stacking order, if a meaningful one exists) to get an accurate stacking list.

Reply:
A complete solution might be more complex indeed.
I only care for the focused window to be the last on the list.
Other than that, I can code something that is simple for the rest of the windows, if there is a good suggestion.
But I am not sure that I am up to coding an algorithm that reconstructs stacking order from the rectangles.

Reply:
> But I actually do not know if up or down is related to the visual stacking.

I don't think this is relevant in the general case, except if some layout stacks. I don't see why we should go through the trouble of implementing this accurately. I even think the previous implementation (trivially) fulfills the spec because we can assume the layouts don't stack. That said, this is technically a breaking change (not that I think that's problematic). 

I'm not familiar with the spec by any chance, it's just that the change seemed unjustified to me.

This was discussed previously in https://github.com/xmonad/xmonad-contrib/pull/568 and https://github.com/xmonad/xmonad-contrib/issues/567. Any thoughts @liskin  @slotThe ?

Reply:
I might have made a mistake, changing the order of the windows within a given workspace span.
My main focus was making sure that the currently focused window is the last in the list.
After reading both #568 and #567, it seems it is a consensus that it should be the case.
All the other of my changes are, hopefully, readability improvements.

I can change the PR to only affect the order of workspace spans within `_NET_CLIENT_LIST_STACKING`.

In my defense, it seems I am not the only one thinking that the current ordering is reversed.
I think this comment also suggested:

```haskell
stackWindows (W.Stack cur up down) = up ++ reverse down ++ [cur]
```
https://github.com/xmonad/xmonad-contrib/issues/567#issuecomment-870855824

And I've made this change, thinking it was just a typo.

Reply:
I think the order is unpredictable to the point it simply does not and can not matter. If I'm reading the code in X.L.Magnifier correctly, in addition to making sure the magnified window is on top, it reverses the order of windowrects that came from the underlying layout:
https://github.com/xmonad/xmonad-contrib/blob/3d65a6bf72236a0b54b63b6c2c71854eb17ed3d3/XMonad/Layout/Magnifier.hs#L226-L231

```haskell
> let mag w ws | w == 3 = ws ++ [w] | otherwise = w : ws in reverse (foldr mag [] [1,2,3,4,5])
[3,5,4,2,1]
```

So we have various layouts returning windowrects in various orders, and then we have modifiers that just reverse the order when used… It's simply not worth trying to find a "best" order here. :-)

Reply:
I've restored the order within a workspace as it was before.
Now the only functional change should be in how the workspace spans are ordered.
The current workspace will be the last, and the currently focused window will be the last as well.

Added a link to this and the other two discussion in a comment.
Could help the next person looking at this code, I think.